app/feature/travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/data/CreateTravelCertificateUseCase.kt

```diff
@@ -24,30 +36,51 @@ class CreateTravelCertificateUseCase(
     isMemberIncluded: Boolean,
     coInsured: List<CoInsured>,
     email: String,
-  ): Either<ErrorMessage, TravelCertificateUrl> = either {
-    val input = TravelCertificateCreateInput(
-      contractId = contractId,
-      startDate = startDate,
-      isMemberIncluded = isMemberIncluded,
-      coInsured = coInsured.map { TravelCertificateCreateCoInsured(it.name, it.ssn) },
-      email = email,
-    )
-
-    val query = TravelCertificateCreateMutation(input)
-
-    val url = apolloClient
-      .mutation(query)
-      .safeExecute()
-      .toEither(::ErrorMessage)
-      .onLeft { e { it.message ?: "Could not create travel certificate" } }
-      .bind()
-      .travelCertificateCreate
-      .signedUrl
-
-    TravelCertificateUrl(url)
+  ): Either<ErrorMessage, TravelCertificateUri> = withContext(Dispatchers.IO) {
+    either {
+      val input = TravelCertificateCreateInput(
+        contractId = contractId,
+        startDate = startDate,
+        isMemberIncluded = isMemberIncluded,
+        coInsured = coInsured.map { TravelCertificateCreateCoInsured(it.name, it.ssn) },
+        email = email,
+      )
+
+      val query = TravelCertificateCreateMutation(input)
+
+      val pdfUrl = apolloClient
+        .mutation(query)
+        .safeExecute()
+        .toEither(::ErrorMessage)
+        .onLeft { e { it.message ?: "Could not create travel certificate" } }
+        .bind()
+        .travelCertificateCreate
+        .signedUrl
+
+      val request = Request.Builder()
+        .url(pdfUrl)
+        .build()
+
+      val downloadedFile = File(
+        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS),
+        CERTIFICATE_NAME,
+      )
+      
+      try {
+        val response = OkHttpClient().newCall(request).await()
+        val buffer = downloadedFile.sink().buffer()
+        buffer.writeAll(response.body!!.source())
+        buffer.close()
+
+        TravelCertificateUri(downloadedFile.absolutePath)
+      } catch (exception: IOException) {
+        e(exception)
+        raise(ErrorMessage("Could not download travel certificate"))
+      }
+    }
```

Comment:
Hmm yeah, that would be quite a lot more involved than this I think.
It'd probably have to involve WorkManager with [an expedited job](https://developer.android.com/guide/background/persistent/getting-started/define-work#expedited) or [android.app.DownloadManager](https://developer.android.com/reference/android/app/DownloadManager) (Didn't find better docs 😅) so that it will show a notification which will in turn show that it's currently downloading.
I tried to look around for some good documentation on this but couldn't find something yet. But did find a bunch of articles that are very old and look weird. Might have to look a bit more to find something better.

Reply:
Check new PR ;)